File: packages/front/src/lib/contracts/dto/arbitrage.ts
import type {
  ArbitrageOpportunityStatus,
  ArbitrageOpportunityType,
  ArbitrageValidationStatus,
} from "./primitives";

export type ArbitrageScanDto = {
  id: string;
  started_at: string;
  finished_at?: string | null;
  market_count: number;
  snapshot_count: number;
  opportunity_count: number;
  scanner_version: string;
  metadata: unknown;
  trace_id: string;
};

export type ArbitrageOpportunityDto = {
  id: string;
  scan_id: string;
  market_id: string;
  opportunity_type: ArbitrageOpportunityType;
  status: ArbitrageOpportunityStatus;
  gross_edge: string;
  price_sum: string;
  capacity: string;
  yes_price: string;
  no_price: string;
  yes_size: string;
  no_size: string;
  observed_at: string;
  reason_codes: string[];
  analysis_payload: unknown;
  trace_id: string;
  validation?: ArbitrageOpportunityValidationDto | null;
};

export type ArbitrageOpportunityValidationDto = {
  id: string;
  opportunity_id: string;
  status: ArbitrageValidationStatus;
  gross_edge: string;
  net_edge: string;
  fee_estimate: string;
  slippage_buffer: string;
  validated_capacity: string;
  book_age_ms: number;
  reason_codes: string[];
  validation_payload: unknown;
  validated_at: string;
  trace_id: string;
};

export type ArbitrageAnalysisRunDto = {
  id: string;
  generated_at: string;
  lookback_hours: number;
  opportunity_count: number;
  market_count: number;
  summary_payload: ArbitrageAnalysisSummaryDto | unknown;
  trace_id: string;
};

export type ArbitrageAnalysisSummaryDto = {
  generated_at: string;
  lookback_hours: number;
  opportunity_count: number;
  market_count: number;
  type_counts: ArbitrageTypeCountDto[];
  top_markets: ArbitrageMarketSummaryDto[];
};

export type ArbitrageTypeCountDto = {
  opportunity_type: ArbitrageOpportunityType;
  count: number;
};

export type ArbitrageMarketSummaryDto = {
  market_id: string;
  opportunity_count: number;
  first_observed_at: string;
  last_observed_at: string;
  duration_seconds: number;
  max_gross_edge: string;
  avg_gross_edge: string;
  max_capacity: string;
  avg_capacity: string;
};

File: packages/backend/crates/contracts/src/dto/arbitrage.rs
// Arbitrage and probability-estimate DTOs: scans, opportunities, validations, analysis runs.

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProbabilityEstimateData {
    pub id: String,
    pub market_id: String,
    pub event_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signal_id: Option<String>,
    pub prior_price: Probability,
    pub posterior_price: Probability,
    pub fair_price: Probability,
    pub market_price: Probability,
    pub edge: Edge,
    pub confidence: Probability,
    pub time_horizon: TimeHorizon,
    pub model_version: String,
    pub reason_codes: Vec<String>,
    pub evidence_count: u32,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArbitrageScanData {
    pub id: String,
    #[serde(with = "time::serde::rfc3339")]
    pub started_at: OffsetDateTime,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[serde(with = "time::serde::rfc3339::option")]
    pub finished_at: Option<OffsetDateTime>,
    pub market_count: u32,
    pub snapshot_count: u32,
    pub opportunity_count: u32,
    pub scanner_version: String,
    pub metadata: Value,
    pub trace_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArbitrageOpportunityData {
    pub id: String,
    pub scan_id: String,
    pub market_id: String,
    pub opportunity_type: String,
    pub status: String,
    pub gross_edge: Edge,
    pub price_sum: String,
    pub capacity: Quantity,
    pub yes_price: Probability,
    pub no_price: Probability,
    pub yes_size: Quantity,
    pub no_size: Quantity,
    #[serde(with = "time::serde::rfc3339")]
    pub observed_at: OffsetDateTime,
    pub reason_codes: Vec<String>,
    pub analysis_payload: Value,
    pub trace_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub validation: Option<ArbitrageOpportunityValidationData>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArbitrageOpportunityValidationData {
    pub id: String,
    pub opportunity_id: String,
    pub status: String,
    pub gross_edge: Edge,
    pub net_edge: Edge,
    pub fee_estimate: Edge,
    pub slippage_buffer: Edge,
    pub validated_capacity: Quantity,
    pub book_age_ms: u64,
    pub reason_codes: Vec<String>,
    pub validation_payload: Value,
    #[serde(with = "time::serde::rfc3339")]
    pub validated_at: OffsetDateTime,
    pub trace_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArbitrageAnalysisRunData {
    pub id: String,
    #[serde(with = "time::serde::rfc3339")]
    pub generated_at: OffsetDateTime,
    pub lookback_hours: u16,
    pub opportunity_count: u32,
    pub market_count: u32,
    pub summary_payload: Value,
    pub trace_id: String,
}
